MIDAS/src/hardware/telemetry.cpp

```diff
@@ -0,0 +1,425 @@
+/**
+ * @file telemetry.cpp
+ *
+ * @brief This file defines the telemetry class used to facilitate
+ * telemetry commands and data transfer between the on-board flight
+ * computer and the ground station.
+ *
+ * Spaceshot Avionics 2023-24
+ * Illinois Space Society - Software Team
+ * Gautam Dayal
+ * Nicholas Phillips
+ * Patrick Marschoun
+ * Peter Giannetos
+ * Rishi Gokkumutkkala
+ * Aaditya Voruganti
+ */
+#include <limits>
+
+#include "telemetry.h"
+#include "../rocket_state.h"
+//#include "rocket_state.h"
+//#include "mcu_main/debug.h"
+//#include "mcu_main/dataLog.h"
+//#include "Rt.h"
+#include <cstring>
+#include <cmath>
+
+#ifndef ENABLE_SILSIM_MODE
+#include <RHHardwareSPI.h>
+#endif
+
+/**
+ * @brief This function maps an input value onto within a particular range into a fixed point value of a certin binary
+ * size
+ *
+ * @param val: number to map into target range, values outside of the range will be clamped
+ *
+ * @param range: range to map number into. For unsigned output, [0, range). For signed output [-range/2, range)
+ *
+ * @return fixed point value represing val mapped onto the target range
+ */
+template <typename T>
+T inv_convert_range(float val, float range) {
+    size_t numeric_range = (int64_t)std::numeric_limits<T>::max() - (int64_t)std::numeric_limits<T>::min() + 1;
+    float converted = val * (float)numeric_range / range;
+    return std::max(std::min((float)std::numeric_limits<T>::max(), converted), (float)std::numeric_limits<T>::min());
+}
+
+ErrorCode Telemetry::init() {
+#if defined(ENABLE_TELEMETRY) && !defined(ENABLE_SILSIM_MODE) 
```

Comment:
make sure to test with ENABLE_TELEMETRY on, it doesn't compile as is

Reply:
there are a lot of errors that show up if you don't #define ENABLE_TELEMETRY